Make deprecated `IO.onError` package private

CJSmith-0141 opened the pull request:
Also adds a test to make sure that code outside the package can still compile with a total function. 

Closes #4329 

Thank you @armanbilge for the solve and @Grogs for the writeup.

CJSmith-0141 committed:
Make deprecated `IO.onError` package private

CJSmith-0141 commented:
Looks like making it package private doesn't work for `2.12`. Maybe un deprecate instead?

joroKr21 commented:
I think on 2.12 you can't use a total function in place of a partial function. So that test would have to be only for 2.13+

CJSmith-0141 commented:
Ok since this would be a source breaking change for Scala `2.12`, I changed the base branch for this PR to be `series/3.x`.

I also moved the test to be 2.13+.

joroKr21 commented:
The previous change was already source breaking because Scala 2.12 can't infer the parameter type of the function passed to an overloaded method like `onError { case e => ... }` will complain about unknown function parameter type.

armanbilge reviewed (approved):
thanks!

djspiewak commented:
@armanbilge Are we going to retarget this to 3.6?

armanbilge reviewed (changes requested):
> Are we going to retarget this to 3.6?

Yes!

armanbilge commented:
superseded https://github.com/typelevel/cats-effect/pull/4343
